Introduce IndexReshardingMetadata

bcully opened the pull request:
This adds to IndexMetadata the persistent state we will need to track while a split is in progress. Nothing outside test code sets it yet, so it doesn't introduce any wire changes yet.

Followups will consult this to make routing decisions and handle backward compatibility if the object is present in metadata.

bcully committed:
Introduce IndexReshardingMetadata

This adds to IndexMetadata the persistent state
we will need to track while a split is in progress.
Nothing outside test code sets it yet, so it doesn't
introduce any wire changes yet.

Followups will consult this to make routing decisions and
handle backward compatibility if the object is present
in metadata.

elasticsearchmachine commented:
Pinging @elastic/es-distributed-indexing (Team:Distributed Indexing)

Tim-Brooks commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,315 @@
+
+    public IndexReshardingMetadata(StreamInput in) throws IOException {
+        this(
+            in.readInt(),
```
Not that it matters that much, but I don't think we need to store this if we are also storing an array of sources and targets. We can just read the array size.

bcully replied:
That's true. An earlier revision wasn't tracking source shard state, so I needed it. I'm kind of tempted to keep it just for clarity, but I agree it's redundant.

Tim-Brooks replied:
In my opinion it is fine to keep it as a dedicated field in memory, however, we should not serialize additional data in the cluster state that is not needed.

bcully replied:
ok

bcully replied:
Done in 7f6c7b9.

Tim-Brooks commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,315 @@
+        }
+    }
+
+    public enum TargetShardState implements Writeable {
```
I wonder if we need the "State" to be more like a Reshard enum (SOURCE (SOURCE - maybe we need to change this name for state / DONE), TARGET (CLONE, HANDOFF, SPLIT, DONE)) itself with variable characteristics on it. Similar to `RecoverySource` . 

And then it is:
- old [SOURCE (state: PRE_DELETE), SOURCE (state: PRE_DELETE)]
- new [TARGET (state: CLONE), TARGET (state: HANDOFF)]

Which then allows for an additional types once we implement shrink that can be serialized into the same old / new arrays?

Just something to think about before we commit to an initial serialization.

bcully replied:
I will think about it for a bit. It'd be nice to avoid a situation where an array could hold both SPLIT(SOURCE) and SHRINK(SOURCE) elements at the same time. I wonder whether we might want to wrap the whole array pair in an operation or something.

bcully replied:
I think we might get the extensibility you want and the type safety I want by adding an "operation" enum (SPLIT, SHRINK) before the shard arrays in the serialized form. On parsing the operation we could branch to parsing the remainder of the serialized metadata according to the operation. I'll give it a try, assuming that works for you?

Tim-Brooks replied:
I don't have a strong opinion about how it is designed. I just suspect:

```
SplitSource
SplitTarget

ShrinkSource
ShrinkTarget
```

Will have different states associated with them so we might think about how we represent that. An enum for the operation type would also provide that (I assume).

bcully replied:
alright, makes sense, I'll play with a top level field to control how the rest of the resharding state is interpreted.

bcully replied:
I've made the IndexReshardingMetadata field into something like a tagged union in 7f6c7b9. To be honest, it felt like I was fighting the XContent model a little bit doing it, so it could be I'm missing something, but it does appear to work. I tried to find other examples of XContent that held versioned data but didn't see much applicable. Thoughts appreciated here.

bcully replied:
I've simplified this in 272e8f3

Tim-Brooks reviewed (commented):
Looks pretty good to begin with. I'm glad you were able to implement the core serialization tests.

bcully committed:
Documentation updates (feedback from ankikuma)

bcully commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,315 @@
+ * is occurring, to ensure that we don't lose or double-count documents during the process. We prevent this
+ * by maintaining the state of the split on the source and target shards, and making an atomic (from the point
+ * of view of indexing and search requests) transition from handling requests that route to the target shard
+ * on the source shard, to letting the target shard handle them.
```
addressed in 790a2d2

bcully commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,315 @@
+ *     its store that are now the responsibility of the target shards and transitions to `DONE`.
+ *   This isn't strictly required to be persistent for correctness, but it can save time on recovery
+ *   by allowing a DONE shard to skip interrogating targets and repeating cleanup.
+ * * For each target shard, its current target state, one of `CLONE`, `HANDOFF`, `SPLIT`, or `DONE`.
```
addressed in 790a2d2

bcully commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,315 @@
+ *     after handoff. So at HANDOFF, the source shards continue to service searches, but block refresh since they cannot
+ *     be guaranteed to have seen documents indexed after HANDOFF.
+ *   - When the target shard's corresponding search replica has started running, the target requests that the source filter
+ *     search results belonging to the target, and moves the target shard's state moves to `SPLIT`. The target's search replica
```
addressed in 790a2d2

bcully commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,315 @@
+ *   - It moves to `HANDOFF` when it has copied all of its data from the source to indicate that it is now ready to
+ *     receive indexing actions, and starts RUNNING. After this point, the source may no longer contain the entire contents
+ *     of the target and must not index documents belonging to the target. But since search shards can't start up until
+ *     their corresponding index nodes are RUNNING, search requests would fail if they routed to the target shard immediately
```
addressed in 790a2d2

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-10565

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-10565

bcully committed:
Merge branch 'main' into ES-10565

bcully committed:
Make IndexReshardingMetadata extensible

This refactors IndexReshardingMetadata to hold a sum type
for persistent state that can be extended with new operations.
For instance we will want to add a shrink operation in the future,
or we may need to create a new version of split later.

Making XContent support polymorphic data felt clunky - perhaps
there is a different approach that I'm missing?

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-10565

bcully committed:
Simplify IndexReshardingMetadata parser

Previously it was using a discriminant field to control how to
parse the rest of the metadata, which it then did by hand:

{ "operation": "split", ... }

Now it uses a separate key for each kind of metadata, and uses
ObjectParser's `declareExclusiveFieldSet` machinery to enforce
that only one exists:

{ "split": { ... } }

The down side is that the state field is no longer final,
because we don't know which kind of state we have at construction
time due to a limitation of the ConstructingObjectParser interface.

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-10565

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,215 @@
+    }
+
+    public IndexReshardingMetadata(StreamInput in) throws IOException {
+        this();
```
I would remove this (in particular once we remove the default constructor/make the field final) since we expect the read here to fill all fields.

bcully replied:
Done in 0e69a76

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,215 @@
+        this();
+        var stateName = in.readString();
+
+        if (stateName.equals(SPLIT_FIELD.getPreferredName())) {
```
I have a slight preference for switch here - just to keep it similar to `writeTo` mostly, but also a bit shorter I think. Really nit-picking, same order would also be good ;-).

bcully replied:
this was an if only because of the `getPreferredName` business (can't write `case SPLIT_FIELD.getPreferredName():`). I'll just create consts for the bare strings as well.

bcully replied:
Done in 0e69a76

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,215 @@
+ * to beginning another resharding operation.
+ */
+public class IndexReshardingMetadata implements ToXContentFragment, Writeable {
+    // ideally this would be final but the parser can't currently set it at construction time
```
I notice you using that in `Split` so maybe I am missing the underlying reason why we cannot construct the metadata with the state passed into the constructor?

bcully replied:
I had tried to use `ConstructingObjectParser` before and do prefer it. I ran into trouble because I was trying to use a single argument constructor (taking the state) populated by either of two different parsers (one for `Split` and one for `Noop`). I must have made an error in testing when I tried out an `optionalConstructorArg` approach last time, because I was able to get it to work just now. Thanks for getting me to revisit it, I'll include this change in the next refresh.

bcully replied:
Done in 0e69a76

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java:
```diff
@@ -0,0 +1,300 @@
+
+            @Override
+            public void writeTo(StreamOutput out) throws IOException {
+                out.writeEnum(this);
```
This relies on the enum ordinal, which is a bit annoying if we ever want to change the enum. I think we prefer to add an explicit number - or use the name. See for instance [this example](https://github.com/elastic/elasticsearch/blob/f2b09051eca48e869ffdde754a6af69b1d3fd0bf/server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java#L980)

bcully replied:
Alright, I'll add an explicit key.

bcully replied:
Done in 0e69a76

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java:
```diff
@@ -0,0 +1,300 @@
+            DONE;
+
+            @Override
+            public void writeTo(StreamOutput out) throws IOException {
```
I think we also want a constructor with `StreamInput` then? Can be follow-up for sure if it is not needed here.

bcully replied:
When I add the explicit number then I'll also add a static `readFrom` as in the example you linked. In this iteration it wasn't valuable since the parent was just calling `readEnum` inside `readArray`.

bcully replied:
Done in 0e69a76

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java:
```diff
@@ -0,0 +1,300 @@
+            return sourceShards;
+        }
+
+        public TargetShardState[] targetShards() {
```
It is not entirely clear to me if this is indexed by shardId or by (shardId-sourceShards)? I.e., we go from 2 to 4 shards, we have source shards 0-1 and target shards 2-3, what is the return value here?

I wonder if we need to expose it as arrays? Perhaps it is better to do `targetShard(shardId)` in the client code?

henningandersen replied:
I see it now in the method below. I think exposing it as an array is somewhat involved to callers?

henningandersen replied:
I also see the getter, so I propose to just remove these or make them more private.

bcully replied:
yes, thanks, these are there just for the serialization tests really. I forgot to adjust their access modifiers, will do that now.

bcully replied:
Done in 0e69a76

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java:
```diff
@@ -0,0 +1,300 @@
+                assert target == TargetShardState.DONE : "can only move source shard to DONE when all targets are DONE";
+            }
+
+            sourceShards[shardNum] = sourceShardState;
```
We should not mutate these objects. This is way too dangerous, since the index-metadata is expected to be immutable. Instead we should add a `Builder` and use this to build a new `IndexReshardingState`.

bcully replied:
Thanks, I agree this is worth doing now. I should admit that I haven't really put a lot of thought into the client API for this state yet, since I thought that would be easier to get right when we were building the consumers. I expect it to churn until we actually have the state machine built into the resharding process.

bcully replied:
Done in 0e69a76 (but the API is still just enough to support a unit test for the moment)

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java:
```diff
@@ -0,0 +1,300 @@
+            assert targetShardNum >= 0 && targetShardNum < targetShards.length : "target shardNum is out of bounds";
+            assert targetShards[targetShardNum].ordinal() + 1 == targetShardState.ordinal() : "invalid target shard state transition";
+
+            targetShards[targetShardNum] = targetShardState;
```
Similar to above, we should make this object immutable.

bcully replied:
will do.

bcully replied:
Done in 0e69a76

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java:
```diff
@@ -0,0 +1,300 @@
+         * @param shardNum a source shard index greater than or equal to 0 and less than the original shard count
+         * @return an array of target shard states in order for the given shard
+         */
+        public TargetShardState[] getTargetStatesFor(int shardNum) {
```
I think we should either return a Collection of such states - or a Map<int, TargetShardState> to make the shard-id explicit.

I wonder if we even need to return this though or if we can add the methods needed to this class instead? Like the `inProgress` method below, but related to a specific source shard.

bcully replied:
I'll make this private and expose a `targetsDone(int sourceShardIdx)` or similar instead for the unit test.

bcully replied:
Done in 0e69a76

henningandersen reviewed (commented):
Read through the non-test code, left a few comments, otherwise looks good.

bcully committed:
IndexReshardingMetadata PR feedback

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-10565

bcully committed:
IndexReshardingMetadata PR feedback

Tim-Brooks commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java:
```diff
@@ -0,0 +1,214 @@
+        return builder;
+    }
+
+    public IndexReshardingMetadata(StreamInput in) throws IOException {
```
We mostly put ctors at the tope of classes, below fields. So I feel like this would best go above `getState`.

bcully replied:
Done in 20264b8

Tim-Brooks commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingState.java:
```diff
@@ -0,0 +1,383 @@
+            return SPLIT_PARSER.parse(parser, null);
+        }
+
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
Override?

bcully replied:
Thanks, done in 20264b8. I'm surprised intellij didn't give me a squiggle for that one.

Tim-Brooks reviewed (approved):
This looks fine to me with mostly minor comments.

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-10565

bcully committed:
IndexReshardingMetadata PR feedback

Move constructors up; mark a toXContent @Override

ankikuma reviewed (approved):
Looks good Brendan!

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-10565
